src/ws/cert-session.c

```diff
@@ -0,0 +1,307 @@
+
+  user_obj_v = g_dbus_connection_call_sync (bus,
+                                            "org.freedesktop.sssd.infopipe",
+                                            "/org/freedesktop/sssd/infopipe/Users",
+                                            "org.freedesktop.sssd.infopipe.Users",
+                                            "FindByCertificate",
+                                            g_variant_new ("(s)", certificate),
+                                            type_object_path_tuple,
+                                            G_DBUS_CALL_FLAGS_NONE,
+                                            -1, /* default timeout */
+                                            NULL,
+                                            &error);
+  if (!user_obj_v)
+    {
+      g_warning ("Failed to map certificate to user: %s", error->message);
+      exit_init_problem (g_error_matches (error, G_DBUS_ERROR, G_DBUS_ERROR_SERVICE_UNKNOWN)
+                           ? PAM_AUTHINFO_UNAVAIL
+                           : PAM_USER_UNKNOWN);
+    }
+
+  g_variant_get (user_obj_v, "(&o)", &user_obj);
+  assert (user_obj);
+
+  username_v = g_dbus_connection_call_sync (bus,
+                                            "org.freedesktop.sssd.infopipe",
+                                            user_obj,
+                                            "org.freedesktop.DBus.Properties",
+                                            "Get",
+                                            g_variant_new ("(ss)", "org.freedesktop.sssd.infopipe.Users.User", "name"),
+                                            NULL,
+                                            G_DBUS_CALL_FLAGS_NONE,
+                                            -1, /* default timeout */
+                                            NULL,
+                                            &error);
+  if (!username_v)
+    {
+      g_warning ("Failed to map user object to name: %s", error->message);
+      exit_init_problem (PAM_SERVICE_ERR);
+    }
+
+  g_variant_get (username_v, "(v)", &username_vv);
+  assert (username_vv);
+  if (username_vv && g_variant_is_of_type (username_vv, G_VARIANT_TYPE_STRING))
+    {
+      username = g_variant_dup_string (username_vv, NULL);
+    }
+  else
+    {
+      g_critical ("Invalid return type of user property: %s", g_variant_get_type_string (username_v));
+      exit_init_problem (PAM_SERVICE_ERR);
+    }
+
+  assert (username);
+  debug ("mapped certificate to user %s", username);
+  return g_strdup (username);
+}
+
+static pam_handle_t *
+perform_tlscert (const char *rhost,
+                 const char *auth_value)
+{
+  struct pam_conv conv = { pam_conv_func, };
+  pam_handle_t *pamh;
+  g_autofree char *certificate = NULL;
+  g_autofree char *user = NULL;
+  int res;
+
+  if (!auth_value || !*auth_value )
+    errx (EX, "empty authentication value");
+
+  debug ("tls-cert authentication with %s", auth_value);
+
+  /* the auth_value has escaped line breaks */
```

Comment:
Lets add a massive TODO here that the certificate should be coming from a secure source, since it's blindly trusted after this point.